Create engine outside

Zelldon opened the pull request:
## Description

Create the engine outside of the StreamProcessor, which should allow to also inject other RecordProcessors.
Furthermore we can fill the Engine with the needed TypedRecordProcessorFactory, which means we no longer need that in our builders and contexts.

I hope this helps you in make further progress @deepthidevaki 


Sidenote:

This also simplifies the StreamProcessor/RecordProcessor tests, since we do not necessarily need to set up so much anymore.
<!-- Please explain the changes you made here. -->

## Related issues

<!-- Which issues are closed by this PR or are related -->

related to https://github.com/camunda/zeebe/issues/9725

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

deepthidevaki reviewed (changes requested):
:rocket: Exactly what I wanted :smile:  Just some minor comment.

Zelldon commented on engine/src/main/java/io/camunda/zeebe/engine/RecordProcessorContext.java:
```diff
@@ -21,7 +20,7 @@
 import java.util.function.Function;
 
-public final class EngineContext {
+public final class RecordProcessorContext {
```
I don't see that as blocker since we are not done yet but sure can do tomorrow.

Zelldon commented:
@deepthidevaki I moved the implementation and introduced a new interface. Could you review again? Maybe you trust me next time :grin: so I can just go ahead and merge after the changes :)

deepthidevaki commented:
> @deepthidevaki I moved the implementation and introduced a new interface. Could you review again? Maybe you trust me next time grin so I can just go ahead and merge after the changes :)

:confused: I didn't know that we now approve and request changes at the same time.

deepthidevaki reviewed (approved):
Thanks.

Zelldon commented:
Since this was only something minor I thought so :) https://github.com/camunda/zeebe/blob/main/CONTRIBUTING.md#reviewing-a-pull-request

Zelldon commented:
bors r+
